### PR TITLE
Fix newnames handling for new local vars

### LINKED ISSUE #3437: Simple Java program with array creation cannot be sliced
## Description

It is not possible to slice the following program when pruning the proof and rerunning automode.

```
class Newnames {
    /*@ normal_behaviour
      @ ensures true;
      @*/
    void createArray() {
        int[] a = new int[5 + 1];
    }
}

```

## Reproducible
  
always

### Steps to reproduce 

1. Load Newnames.java
2. Run automode to completion
3. Prune to node 13
4. Run automode again
5. Try to slice the proof

> What is your expected behavior and what was the actual behavior?
Expected: slicing works
Actual: `failed to find new formula @ variableDeclaration (serial nr 112)`

### Additional information

Stacktrace:
<details>
```
java.lang.IllegalStateException: failed to find new formula @ variableDeclaration (serial nr 112)
	at de.uka.ilkd.key.proof.replay.AbstractProofReplayer.constructTacletApp(AbstractProofReplayer.java:263)
	at de.uka.ilkd.key.proof.replay.AbstractProofReplayer.reApplyRuleApp(AbstractProofReplayer.java:103)
	at org.key_project.slicing.SlicingProofReplayer.slice(SlicingProofReplayer.java:202)
	at org.key_project.slicing.ui.SlicingLeftPanel.lambda$sliceProof$8(SlicingLeftPanel.java:447)
	at org.key_project.slicing.util.GenericWorker.doInBackground(GenericWorker.java:48)
	at org.key_project.slicing.util.GenericWorker.doInBackground(GenericWorker.java:16)
	at java.desktop/javax.swing.SwingWorker$1.call(SwingWorker.java:304)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.desktop/javax.swing.SwingWorker.run(SwingWorker.java:343)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
```
</details>

---

* Commit: 9ad6cc893ef6e61574fac6994511791b8411ccfc

### PR DESCRIPTION
**Pull request by FliegendeWurst**
## Related Issue

This pull request fixes #3437.

## Intended Change

Previously, the names of certain local variables were derived using a counter. Now they will also be stored in the name proposals, making it possible to replay their exact value.

## Type of pull request

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactoring (behaviour should not change or only minimally change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] There are changes to the (Java) code
- [ ] There are changes to the taclet rule base
- [ ] There are changes to the deployment/CI infrastructure (gradle, github, ...)
- [ ] Other: 

## Ensuring quality
    
- [x] I made sure that introduced/changed code is well documented (javadoc and inline comments).
- [ ] I made sure that new/changed end-user features are well documented (https://github.com/KeYProject/key-docs).
- [x] I added new test case(s) for new functionality.
- [x] I have tested the feature as follows: automated test specifically for #3437
- [ ] I have checked that runtime performance has not deteriorated.

## Additional information and contact(s)

The contributions within this pull request are licensed under GPLv2 (only) for inclusion in KeY.

### CODE REVIEW

File: key.core/src/main/java/de/uka/ilkd/key/logic/VariableNamer.java [hunk 1]
```diff
@@ -48,6 +48,8 @@
 public abstract class VariableNamer implements InstantiationProposer {
     private static final Logger LOGGER = LoggerFactory.getLogger(VariableNamer.class);
 
+    public static final char TEMPORARY_VAR_SEPARATOR = '#';
```
Comment (WolframPfeifer) [inline key.core/src/main/java/de/uka/ilkd/key/logic/VariableNamer.java:51]: I see the point that it is necessary to move this constant, but I think the name is misleading. The original one was `TempIndProgramElementName.SEPARATOR`, which indicates that it is used to mark a temporary **index**.
Maybe we should call it "TEMP_INDEX_SEPARATOR" and also provide a JavaDoc here.

**Review by WolframPfeifer — CHANGES_REQUESTED** [review]
I think this PR is fine, apart from my small comment about the constant name.

File: key.core/src/main/java/de/uka/ilkd/key/logic/VariableNamer.java (hunk 1 above)
Comment (FliegendeWurst) [reply key.core/src/main/java/de/uka/ilkd/key/logic/VariableNamer.java]: I have renamed the constant and added some JavaDoc.

**Review by WolframPfeifer — APPROVED** [review]

**Comment by mattulbrich** [comment]
Does this also fix #3455 where a method parameter `length` shadows the `length` function for array length?

**Pull request merged**